--- /Users/kennethsharp1/WWW/d7_ses_maketest/sites/all/modules/stanford_carousel/js/stanford_carousel_aria.js	2015-07-19 16:26:32.000000000 -0700
+++ stanford_carousel_aria.js	2015-07-16 20:27:26.000000000 -0700
@@ -11,8 +11,8 @@
     this.$indicators = this.$element.find('.carousel-indicators')
     this.options = options
     this.options.pause == 'hover' && this.$element
-      .on('mouseenter', $.proxy(this.pause, this))
-      .on('mouseleave', $.proxy(this.cycle, this))
+      //.on('mouseenter', $.proxy(this.pause, this))
+      //.on('mouseleave', $.proxy(this.cycle, this))
   }
 
   Carousel.prototype = {
